sandbox/cgroup/scanning.go

```diff
@@ -0,0 +1,155 @@
+// If the feature wasn't always enabled then only service process are correctly
+// accounted for.
+//
+// The return value is a snapshot of the pids for a given snap, grouped by
+// security tag. The result may be immediately stale as processes fork and
+// exit.
+//
+// Importantly, if the per-snap lock is held while computing the set, then the
+// following guarantee is true: if a security tag is not among the results then
+// no such tag can come into existence while the lock is held.
+//
+// This can be used to classify the activity of a given snap into activity
+// classes, based on the nature of the security tags encountered.
+func PidsOfSnap(snapInstanceName string) (map[string][]int, error) {
+	// pidsByTag maps security tag to a list of pids.
+	pidsByTag := make(map[string][]int)
+	securityTagPrefix := "snap." + snapInstanceName + "."
+
+	// Walk the cgroup tree and look for "cgroup.procs" files. Having found one
+	// we try to derive the snap security tag from it. If successful and the
+	// tag matches the snap we are interested in, we harvest the snapshot of
+	// PIDs that belong to the cgroup and put them into a bucket associated
+	// with the security tag.
+	walkFunc := func(path string, fileInfo os.FileInfo, err error) error {
+		if err != nil {
+			// See the documentation of path/filepath.Walk. The error we get is
+			// the error that was encountered while walking. We just surface
+			// that error quickly.
+			return err
+		}
+		if fileInfo.IsDir() {
+			// We don't care about directories.
+			return nil
+		}
+		if filepath.Base(path) != "cgroup.procs" {
+			// We are looking for "cgroup.procs" files. Those contain the set
+			// of processes that momentarily inhabit a cgroup.
+			return nil
+		}
+		// Now that we are confident that the file we're looking at is
+		// interesting, extract the security tag from the cgroup path and check
+		// if the security tag belongs to the snap we are interested in. Since
+		// not all cgroups are related to snaps it is not an error if the
+		// cgroup path does not denote a snap.
+		cgroupPath := filepath.Dir(path)
+		securityTag := securityTagFromCgroupPath(cgroupPath)
+		if securityTag == "" {
+			return nil
+		}
+		if !strings.HasPrefix(securityTag, securityTagPrefix) {
+			return nil
+		}
+		pids, err := pidsInFile(path)
+		if err != nil {
+			return err
+		}
+		pidsByTag[securityTag] = append(pidsByTag[securityTag], pids...)
+		// Since we've found the file we are looking for (cgroup.procs) we no
+		// longer need to scan the remaining files of this directory.
+		return filepath.SkipDir
+	}
+
+	var cgroupPathToScan string
+	ver, err := Version()
+	if err != nil {
+		return nil, err
+	}
+	if ver == V2 {
+		// In v2 mode scan all of /sys/fs/cgroup as there is no specialization
+		// anymore (each directory represents a hierarchy with equal
+		// capabilities and old split into controllers is gone).
+		cgroupPathToScan = filepath.Join(rootPath, expectedMountPoint)
```

Comment:
this is older than this PR, but I suppose  cgroupExpectedMountPoint would be clearer